src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Snapshotter.cs

```diff
@@ -0,0 +1,250 @@
+                // We cache the last used dictionaries for _variableSlot and variableBySlot,
+                // in order to minimize the number of dictionaries created. As we don't ever
+                // remove or change things in these dictionaries, only add to them, we can just
+                // track whether the underlying data structures have had their sizes changed since
+                // the last time we captured them.
+                // NullableWalker.variableBySlot is allocated in an exponential fashion: every time
+                // a new slot is created that's off the end of the array, the array size is doubled.
+                // Therefore, we only want to take the allocated identifiers, and verify that any slots
+                // after the ones that we have are empty in Debug to verify. Slot 0 is always allocated,
+                // so if _lastVariableBySlot is empty then we've never cached anything and need to cache
+                if (_walkerToSnapshot._variableSlot.Count != _lastVariableSlot.Count
+                    || _lastVariableBySlot.Length == 0)
+                {
+                    var variableBySlotBuilder = ArrayBuilder<VariableIdentifier>.GetInstance(_walkerToSnapshot.variableBySlot.Length);
+                    var walkerVariableBySlot = _walkerToSnapshot.variableBySlot;
+                    int i;
+                    for (i = 0; i < walkerVariableBySlot.Length; i++)
+                    {
+                        // The first slot always does not exist for the purposes of caching and can be ignored
+                        // in the exit condition check
+                        if (i != 0 && !walkerVariableBySlot[i].Exists)
+                        {
+                            break;
+                        }
+
+                        variableBySlotBuilder.Add(walkerVariableBySlot[i]);
+                    }
+
+#if DEBUG
+                    verifyRemainderNonExistant(walkerVariableBySlot, i);
+#endif
+
+                    _lastVariableBySlot = variableBySlotBuilder.ToImmutableAndFree();
+                    _lastVariableSlot = _walkerToSnapshot._variableSlot.ToImmutableDictionary();
+                }
+#if DEBUG
+                else
+                {
+                    // In debug mode, we verify the contents of the arrays to ensure that the invariant
+                    // is maintained. Because variableBySlot is allocated exponentially, we can only
+                    // verify that the length of our cached version is equal or less than the source
+                    // array, and then verify that anything we haven't cached does not exist.
+                    Debug.Assert(_lastVariableBySlot.Length <= _walkerToSnapshot.variableBySlot.Length);
+                    int i;
+                    for (i = 0; i < _lastVariableBySlot.Length; i++)
+                    {
+                        Debug.Assert(compareIdentifiers(_lastVariableBySlot[i], _walkerToSnapshot.variableBySlot[i]));
```

Comment:
>Debug.Assert(compareIdentifiers(_lastVariableBySlot[i], _walkerToSnapshot.variableBySlot[i])); [](start = 24, length = 94)

It looks like this code can be simplified if we assert that ```_lastVariableBySlot.Length == nextVariableSlot```

Reply:
Assuming that we handle always empty 0's element outside of the loop.

---
In reply to: [288757319](https://github.com/dotnet/roslyn/pull/35850#discussion_r288757319) [](ancestors = 288757319)